File: src/api/subsonic/methods/getTopSongs.ts
import * as v from 'valibot';

import {
  handleJsonResponse,
  makeSubsonicRequest,
  parseJsonResponse,
} from '../makeRequest';
import { Song } from '../types/song';

export function subsonicGetTopSongs(
  artist: string,
  { count }: { count?: number } = {},
) {
  return makeSubsonicRequest({ method: 'rest/getTopSongs', artist, count })
    .flatMap(response =>
      parseJsonResponse(response, {
        topSongs: v.object({ song: v.optional(v.array(Song)) }),
      }),
    )
    .flatMap(json => handleJsonResponse(json['subsonic-response']))
    .map(result => result.topSongs.song);
}

File: src/api/subsonic/getTopSongs.ml
open Fx
open Monad_syntax
open Utils

let make artistName ?count () =
  let* { response = { status = `ok; topSongs = { song } } } =
    let params = Js.Dict.empty () in
    Js.Dict.set params "artist" (Value artistName);
    count
    |> Option.iter (fun count ->
           Js.Dict.set params "count" (Value (Int.to_string count)));
    makeRequest { _method = "rest/getTopSongs"; params } TopSongs.of_json
  in
  Ok song
